File: backend/app/modules/simulation/clinical_order_schemas.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field, field_validator

from app.modules.clinical_orders.entities import FastingPolicy, RoomServiceType
from app.modules.routing.schemas import RouteProposalResponse, ScheduleStrategy
from app.shared.enums import RoutePriority


class ClinicalOrderStatus(StrEnum):
    ROUTED = "routed"


class DispatchClinicalOrderRequest(BaseModel):
    patient_code: str = Field(min_length=3, max_length=40)
    patient_name: str = Field(min_length=2, max_length=120)
    encounter_id: str = Field(min_length=3, max_length=60)
    doctor_name: str = Field(min_length=2, max_length=120)
    doctor_room_code: str = Field(default="PK-305", min_length=2, max_length=40)
    clinical_service_codes: list[str] = Field(min_length=1, max_length=12)
    priority: RoutePriority = RoutePriority.FASTEST
    schedule_strategy: ScheduleStrategy = ScheduleStrategy.BALANCED

    @field_validator(
        "patient_code",
        "patient_name",
        "encounter_id",
        "doctor_name",
        "doctor_room_code",
        mode="before",
    )
    @classmethod
    def strip_text(cls, value: object) -> object:
        return value.strip() if isinstance(value, str) else value

    @field_validator("clinical_service_codes")
    @classmethod
    def validate_service_codes(cls, value: list[str]) -> list[str]:
        normalized = [code.strip().upper() for code in value if code.strip()]
        if not normalized:
            raise ValueError("Phải có ít nhất một chỉ định.")
        if len(normalized) != len(set(normalized)):
            raise ValueError("Mỗi chỉ định chỉ được gửi một lần.")
        return normalized


class MatchedRoomResponse(BaseModel):
    code: str
    location_code: str
    name: str
    floor: str
    status: str
    waiting_patients: int = Field(ge=0)
    estimated_wait_minutes: int = Field(ge=0)


class DispatchedClinicalOrderItemResponse(BaseModel):
    service_code: str
    service_name: str
    room_service_type: RoomServiceType
    fasting_policy: FastingPolicy
    fasting_hours_min: int | None
    fasting_hours_max: int | None
    notes: str | None
    configured_room_locations: list[str]
    matched_rooms: list[MatchedRoomResponse]


class ClinicalOrderDispatchResponse(BaseModel):
    id: str
    status: ClinicalOrderStatus
    patient_code: str
    patient_name: str
    encounter_id: str
    doctor_name: str
    doctor_room_code: str
    created_at: datetime
    items: list[DispatchedClinicalOrderItemResponse]
    route_proposal: RouteProposalResponse

File: frontend/src/entities/clinical-order/model/clinical-order.schemas.ts
import { z } from 'zod'

const roomServiceTypeSchema = z.enum([
  'blood_test',
  'urine_test',
  'xray',
  'ultrasound',
  'ct_scan',
])

const fastingPolicySchema = z.enum([
  'not_required',
  'required',
  'conditional',
])

const matchedRoomSchema = z.object({
  code: z.string(),
  location_code: z.string(),
  name: z.string(),
  floor: z.string(),
  status: z.string(),
  waiting_patients: z.number().int().nonnegative(),
  estimated_wait_minutes: z.number().int().nonnegative(),
})

const dispatchedOrderItemSchema = z.object({
  service_code: z.string(),
  service_name: z.string(),
  room_service_type: roomServiceTypeSchema,
  fasting_policy: fastingPolicySchema,
  fasting_hours_min: z.number().int().nonnegative().nullable(),
  fasting_hours_max: z.number().int().nonnegative().nullable(),
  notes: z.string().nullable(),
  configured_room_locations: z.array(z.string()),
  matched_rooms: z.array(matchedRoomSchema),
})

const routeStepSchema = z.object({
  id: z.string(),
  order: z.number().int().positive(),
  service_code: z.string(),
  service_name: z.string(),
  room_code: z.string(),
  room_name: z.string(),
  floor: z.string(),
  wait_minutes_min: z.number().int().nonnegative(),
  wait_minutes_max: z.number().int().nonnegative(),
  service_minutes: z.number().int().positive(),
  travel_minutes: z.number().int().nonnegative(),
  arrival_minutes: z.number().int().nonnegative(),
  complete_minutes: z.number().int().nonnegative(),
  result_ready_minutes: z.number().int().nonnegative(),
  is_locked: z.boolean(),
  lock_reason: z.string().nullable(),
})

const routeOptionSchema = z.object({
  id: z.string(),
  label: z.enum(['recommended', 'less_walk', 'less_crowd']),
  duration_minutes_min: z.number().int().nonnegative(),
  duration_minutes_max: z.number().int().nonnegative(),
  distance_meters: z.number().int().nonnegative(),
  floor_changes: z.number().int().nonnegative(),
  total_wait_minutes: z.number().int().nonnegative(),
  reason: z.string(),
  steps: z.array(routeStepSchema),
})

const routeProposalSchema = z.object({
  id: z.string(),
  encounter_id: z.string(),
  algorithm_version: z.string(),
  updated_at: z.string(),
  expires_at: z.string(),
  warnings: z.array(z.string()),
  options: z.array(routeOptionSchema),
})

export const clinicalOrderDispatchSchema = z.object({
  id: z.string(),
  status: z.literal('routed'),
  patient_code: z.string(),
  patient_name: z.string(),
  encounter_id: z.string(),
  doctor_name: z.string(),
  doctor_room_code: z.string(),
  created_at: z.string(),
  items: z.array(dispatchedOrderItemSchema),
  route_proposal: routeProposalSchema,
})

export type ClinicalOrderDispatch = z.infer<typeof clinicalOrderDispatchSchema>
export type ClinicalOrderItem = ClinicalOrderDispatch['items'][number]

export interface DispatchClinicalOrderPayload {
  patient_code: string
  patient_name: string
  encounter_id: string
  doctor_name: string
  doctor_room_code: string
  clinical_service_codes: string[]
  priority: 'system' | 'fastest' | 'less_walk' | 'less_crowd' | 'accessible'
  schedule_strategy: 'balanced' | 'finish_early' | 'leave_fast'
}
